Die copy & paste, die

mcarton opened the pull request:
This adds two lints to remove copy and paste errors.
~~According to reports [here](http://www.viva64.com/en/examples/V517/) and [there](http://www.viva64.com/en/examples/V523/), those are very common errors.
I took the liberty of defining a “pure” function as a function that has no *&mut* parameter. According to [a comment](https://github.com/Manishearth/rust-clippy/blob/c9342d01213ca1663d2cdf23289bae024823ae6a/src/eq_op.rs#L11):~~
>  ~~False negatives: 
We had some false positives regarding calls (notably [racer](https://github.com/phildawes/racer) had one instance of `x.pop() && x.pop()`), so we removed matching any function or method calls. We may introduce a whitelist of known pure functions in the future.~~

~~function and method calls were removed from `is_exp_equal` because of false positives. This also prevent those particular false positive. Calls to functions with global mutable state will now be considered equal, but in the context where `is_exp_equal` is used, it’d be smelly code anyway (who likes code like `foo() == foo()` or `if foo() { … } else if foo() { … }` where foo modifies some global mutable state?).~~

Ref #588 for the idea. Fix #80.

llogiq commented:
I remember that one. However, this can still fail in the face of interior mutability, right?

mcarton commented:
Ah right, I did not think of that case.
This needs discussion then, would you prefer never to warn things like `foo(bar) == foo(bar)` just in case or leave it at it is and consider relying on hidden mutability here is bad taste anyway?
And how often are those kind of code an error vs. something voluntary?

[Semi-relevant recent comic](http://www.commitstrip.com/en/2016/01/26/genius-or-stupid/).

llogiq commented:
I think we may well want to lint this, but perhaps `allow` by default. People are more forgiving of false positives when they have been warned they can appear (and in which circumstances), and have chosen to activate the lint anyway.

mcarton commented:
PVS-Studio does warn in *Notepad++* for cases like
```C
  } else if (sc.Match('-', '-')) {    //<=
    sc.SetState(SCE_VHDL_COMMENT);
    sc.Forward();
  } else if (sc.Match('-', '-')) {    //<=
```
but I don’t know if the tool is smart enough to know if the method is pure.
Anyway, maybe I can just remove functions again.
I’ll have to improve the PR on Monday anyway, I also forgot to add other expressions (like loops) which might be needed for the `if_same_then_else`.

mcarton commented:
I removed the “naive purity” thing, rebased and completed the `is_exp_equal` function.
Also added `-` and `/` to `EQ_OP` because they [should be there](http://www.viva64.com/en/examples/V501/).

mcarton commented:
Any comment?

Manishearth commented on src/copies.rs:
```diff
@@ -0,0 +1,107 @@
+            false
+        };
+
+        if must_lint {
```
Can't this if be hoisted up as `if is_block_equal`?

(catching such bugs is an interesting lint idea btw)

mcarton replied:
Done. It was probably the result of some refactoring.

Manishearth commented:
LGTM modulo nit

llogiq commented:
Thanks!
